ide/editor.lib2/src/org/netbeans/modules/editor/lib2/view/HighlightsViewUtils.java

```diff
@@ -777,4 +756,85 @@ static View breakView(int axis, int breakPartStartOffset, float x, float len,
+     * if {@code allowWhitespaceBeyondEnd} is true, allows one whitespace character to extend beyond
+     * the preferred break width to make use of all available horizontal space. Very long
+     * unbreakable words may extend beyond the preferred break offset regardless of the setting of
+     * {@code allowWhitespaceBeyondEnd}.
+     *
+     * <p>It was previously considered to allow an arbitrary number of whitespace characters to
+     * trail off the end of each wrap line, rather than just one. In the end, it turned out to be
+     * better to limit this to just one character, as this conveniently avoids the need to ever
+     * position the visual text caret outside the word-wrapped editor viewport (except in cases of
+     * very long unbreakable words).
+     *
+     * @param paragraph a long line of text to be broken, i.e. a paragraph, or the remainder of a
+     *        paragraph if some of its initial lines of wrapped text have already been laid out
+     * @param preferredMaximumBreakOffset the preferred maximum break offset
+     * @param allowWhitespaceBeyondEnd if true, allow one whitespace character to extend beyond
+     *        {@code preferredMaximumBreakOffset} even when this could be avoided by choosing a
+     *        smaller break offset
+     */
+    static int adjustBreakOffsetToWord(CharSequence paragraph,
+            final int preferredMaximumBreakOffset, boolean allowWhitespaceBeyondEnd)
+    {
+        if (preferredMaximumBreakOffset < 0)
+            throw new IllegalArgumentException();
+        if (preferredMaximumBreakOffset > paragraph.length())
+            throw new IllegalArgumentException();
+        /* BreakIterator.getLineInstance already seems to have a cache; creating a new instance here
+        is just the cost of BreakIterator.clone(). So don't bother trying to cache the BreakIterator
+        here. */
+        BreakIterator bi = BreakIterator.getLineInstance(Locale.US);
+        /* Use CharSequenceCharacterIterator to avoid copying the entire paragraph string every
+        time. */
+        bi.setText(new CharSequenceCharacterIterator(paragraph));
+
+        int ret;
+        if (preferredMaximumBreakOffset == 0) {
+            // Skip forward to next boundary.
+            ret = 0;
+        } else if (
+            allowWhitespaceBeyondEnd && preferredMaximumBreakOffset < paragraph.length() &&
+            Character.isWhitespace(paragraph.charAt(preferredMaximumBreakOffset)))
+        {
+            // Allow one whitespace character to extend beyond the preferred break offset.
+            return preferredMaximumBreakOffset + 1;
+        } else {
+            // Skip backwards to previous boundary.
+            ret = bi.isBoundary(preferredMaximumBreakOffset)
+                ? preferredMaximumBreakOffset
+                : bi.preceding(preferredMaximumBreakOffset);
+            if (ret == BreakIterator.DONE)
+                ret = preferredMaximumBreakOffset;
```

Comment:
Use {} around if-ed statement; maybe return `preferredMaximumBreakOffset` immediately for better clarity (case `preferredMaximumBreakOffset == 0` handled at 801).

If `bi.preceding(...) == 0`, that is boundary at char 0, is it OK to search forwards (as seen below) ?

Reply:
OK, fixed cosmetic issues.

Yes, that case is intentional. When the break line is in the middle of a word, we normally scan backwards to try to find the beginning of that word, and then we break there. But if the beginning of the word is in fact at the beginning of the current break line, then that means we have a super-long word that spans the entire viewport, and we must "give up" and break _after_ the word instead, beyond the preferred break offset (requiring horizontal scrolling in the editor). This case is exercised many times when AdjustBreakOffsetToWordTest is run (I just checked to be sure).